flink-connector/src/main/java/com/google/pubsub/flink/internal/source/reader/PubSubNotifyingPullSubscriber.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.pubsub.flink.internal.source.reader;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.ApiService.Listener;
+import com.google.api.core.ApiService.State;
+import com.google.api.core.SettableApiFuture;
+import com.google.cloud.pubsub.v1.AckReplyConsumer;
+import com.google.cloud.pubsub.v1.MessageReceiver;
+import com.google.cloud.pubsub.v1.Subscriber;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import com.google.pubsub.v1.PubsubMessage;
+import java.util.Optional;
+import java.util.concurrent.LinkedBlockingQueue;
+
+public class PubSubNotifyingPullSubscriber implements NotifyingPullSubscriber {
+  public class SubscriberWakeupException extends Exception {}
+
+  public interface SubscriberFactory {
+    Subscriber create(MessageReceiver receiver);
+  }
+
+  private final Subscriber subscriber;
+
+  @GuardedBy("this")
+  private Optional<Throwable> permanentError = Optional.empty();
+
+  @GuardedBy("this")
+  private Optional<SettableApiFuture<Void>> notification = Optional.empty();
+
+  @GuardedBy("this")
+  private final LinkedBlockingQueue<PubsubMessage> messages = new LinkedBlockingQueue<>();
```

Comment:
since its guarded you don't need a concurrent collection, just use an ArrayDeque

Reply:
Done